Foundation/testsuite/src/ProcessRunnerTest.cpp

```diff
@@ -0,0 +1,332 @@
+// and Contributors.
+//
+// SPDX-License-Identifier:	BSL-1.0
+//
+
+
+#include "ProcessRunnerTest.h"
+#include "CppUnit/TestCaller.h"
+#include "CppUnit/TestSuite.h"
+#include "Poco/PIDFile.h"
+#include <sstream>
+#include "Poco/Util/PropertyFileConfiguration.h"
+#include "Poco/Format.h"
+#include "Poco/Path.h"
+#include "Poco/File.h"
+#include "Poco/FileStream.h"
+#include "Poco/Environment.h"
+#include <fstream>
+#include <iostream>
+
+
+//using namespace IoT::Devs::Util;
+using namespace Poco;
+using namespace Poco::Util;
+//using IoT::Devs::Util::enableLogging;
+
+
+ProcessRunnerTest::ProcessRunnerTest(const std::string& name): 
+	CppUnit::TestCase(name)
+{
+	//enableLogging("debug", "%Y-%m-%d %H:%M:%S.%i [%s]: %t");
+}
+
+
+ProcessRunnerTest::~ProcessRunnerTest()
+{
+}
+
+
+void ProcessRunnerTest::testPIDFile()
+{
+	std::string pidFile = Path::tempHome() + "test.pid";
+	std::string pidFiled = Path::tempHome() + "testd.pid";
+
+	{
+		PIDFile f;
+		assertTrue (f.getName().empty());
+		assertTrue (f.getPID() == PIDFile::INVALID_PID);
+		assertFalse (File(pidFile).exists());
+		assertFalse (File(pidFiled).exists());
+
+		f.setName(pidFile, false);
+		assertTrue (f.getName() == pidFile);
+		assertTrue (f.getPID() != PIDFile::INVALID_PID);
+		assertTrue (File(pidFile).exists());
+		assertTrue (!File(pidFiled).exists());
+
+#ifdef _DEBUG
+		f.setName(pidFile, true);
+		assertTrue (f.getName() == pidFiled);
+		assertTrue (f.getPID() != PIDFile::INVALID_PID);
+		assertTrue (!File(pidFile).exists());
+		assertTrue (File(pidFiled).exists());
+#endif
+	}
+	assertFalse (File(pidFile).exists());
+	assertFalse (File(pidFiled).exists());
+
+	{
+		PIDFile f(pidFile);
+#ifdef _DEBUG
+		std::string pf = pidFiled;
+#else
+		std::string pf = pidFile;
+#endif
+
+		assertTrue (f.getName() == pf);
+		assertTrue (File(pf).exists());
+		assertTrue (f.getPID() != PIDFile::INVALID_PID);
+
+		assertTrue (f.exists());
+	}
+	assertFalse (File(pidFile).exists());
+	assertFalse (File(pidFiled).exists());
+
+	{
+		PIDFile f(pidFile, false);
+		assertTrue (f.getName() == pidFile);
+		assertTrue (File(pidFile).exists());
+		assertTrue (f.getPID() != PIDFile::INVALID_PID);
+
+		assertTrue (f.exists());
+	}
+	assertFalse (File(pidFile).exists());
+	assertFalse (File(pidFiled).exists());
+
+	{
+		PIDFile f(pidFile, true, false);
+#ifdef _DEBUG
+		std::string pf = pidFiled;
+#else
+		std::string pf = pidFile;
+#endif
+
+		assertTrue (f.getName() == pf);
+		assertTrue (!File(pf).exists());
+		assertTrue (f.getPID() == PIDFile::INVALID_PID);
+
+		f.create(false);
+		assertTrue (f.exists());
+	}
+	assertFalse (File(pidFile).exists());
+	assertFalse (File(pidFiled).exists());
+}
+
+
+void ProcessRunnerTest::testProcessRunner()
+{
+	std::string name("TestApp");
+	std::string cmd;
+	
+#ifdef _DEBUG
+	name += 'd';
+#endif
+
+	cmd = name;
+
+	// non-auto start, no PID
+	{
+		std::vector<std::string> args;
+		ProcessRunner pr(cmd, args, "", ProcessRunner::NO_OUT, 10, false);
```

Comment:
This works randomly, mostly not. The reason is that you send no command line arguments, the `TestApp` does not wait, completes before `ProcessRunner` can see a valid handle (remember, the `ProcessRunner::run()` is another thread), leaving it trying until it times out and the show is over.

Reply:
Should be fixed in https://github.com/pocoproject/poco/pull/4225/commits/f417b4c8981ad9886383befeb82cae4c18b2430b